grommet/grommet
# [TS] Add TypeScript unit testing support

#### What does this PR do?

This is a proof-of-concept PR related to #5111 that adds _very strict_ TypeScript unit testing support, without forcing the rest of the `grommet` code base to use strict type checks. This means that we can gradually convert existing unit tests from `*-test.js` to `*-test.tsx` files, and fix any issues that TypeScript raises in the process. This should allow us to find and fix existing TypeScript issues quicker, as well as decrease opportunities for TS-related bugs in the future.

#### Where should the reviewer start?

The most important change is in `jest.config.js`, where I specify we want to use the new `ts-jest` dependency for all `tsx?` tests, but use existing pipelines for all `.jsx?` tests.

Secondly, there have been some changes to the `tsconfig.json` file, to support the creation of a more strict `tsconfig.tests.json` file that only applies to files in `**/__tests__` directories. The removed fields are to allow the `string: true` field in the `tsconfig.tests.json` file, and should all be consistent `tsconfig` [defaults](https://www.typescriptlang.org/tsconfig). The reviewer should check that all removed fields are consistent with `tsconfig` [defaults](https://www.typescriptlang.org/tsconfig). 

#### What testing has been done on this PR?

All existing unit tests still pass despite the newly added TypeScript support, both in my local `pre-commit` hook and on CI. See manual testing below for more details.

#### How should this be manually tested?

The reviewer should test that converting a `.js` test file for a few components to `.tsx` should cause tests to fail, since a bunch of new strict checks have been put in place. Failures may range from implicit `any` types through to grommet bugs caused my missing/incorrect prop types. In my testing, I used the `Accordion` and `DataTable` components, however I'd suggest the reviewer tries a few other components too. You can also change some correct prop interfaces to check that errors are correctly identified both in your code editor and when running `yarn test`.

#### Any background context you want to provide?

All context is described here: #5111

#### Do the grommet docs need to be updated? Should this PR be mentioned in the release notes?

No, it is internal to Grommet.

#### Is this change backwards compatible or is it a breaking change?

Backwards compatible
